src/services/filesync/filesync.ts

```diff
@@ -325,97 +322,117 @@ export class FileSync {
+    }
+
+    const { filesVersionHashes, localHashes, gadgetHashes, gadgetFilesVersion } = await this._getHashes();
+    this.log.debug("syncing", { filesVersionHashes, localHashes, gadgetHashes, gadgetFilesVersion });
+
+    if (isEqualHashes(localHashes, gadgetHashes)) {
+      this.log.info("filesystem is in sync");
+      await this._save(gadgetFilesVersion);
+      return;
+    }
+
+    let localChanges = getChanges({ from: filesVersionHashes, to: localHashes, ignore: [".gadget/"] });
+    let gadgetChanges = getChanges({ from: filesVersionHashes, to: gadgetHashes });
+
+    if (localChanges.size === 0 && gadgetChanges.size === 0) {
+      // the local filesystem is missing .gadget/ files
+      gadgetChanges = getChanges({ from: localHashes, to: gadgetHashes });
+      assertAllGadgetFiles({ gadgetChanges });
+    }
+
+    const conflicts = getConflicts({ localChanges, gadgetChanges });
+    if (conflicts.size > 0) {
+      this.log.debug("conflicts detected", { conflicts });
+      printConflicts({
+        message: sprint`{bold You have conflicting changes with Gadget}`,
+        conflicts,
+      });
+
+      const preference = await select({
+        message: "How would you like to resolve these conflicts?",
+        choices: Object.values(ConflictPreference),
+      });
 
-        const stats = await fs.stat(this.directory.absolute(normalizedPath));
-        if (stats.mtime.getTime() > this.mtime) {
-          this.log.trace("detected local change", { path: normalizedPath, mtime: stats.mtime.getTime() });
-          changes.set(normalizedPath, { type: "update" });
+      switch (preference) {
+        case ConflictPreference.CANCEL: {
+          process.exit(0);
+          break;
+        }
+        case ConflictPreference.LOCAL: {
+          gadgetChanges = withoutConflictingChanges({ conflicts, changes: gadgetChanges });
+          break;
+        }
+        case ConflictPreference.GADGET: {
+          localChanges = withoutConflictingChanges({ conflicts, changes: localChanges });
+          break;
         }
       }
-      return changes;
-    };
-
-    let localChanges = await getLocalChanges();
-    const hasLocalChanges = localChanges.size > 0;
-    if (hasLocalChanges) {
-      printChanges({
-        changes: localChanges,
-        tense: "past",
-        message: sprint`{bold Local files have changed since you last synced}`,
-        spaceY: 1,
-        limit: Infinity,
-      });
     }
 
-    let action: Action | undefined;
-    if (hasLocalChanges) {
-      action = await select({
-        message: hasRemoteChanges ? "Remote files have also changed. How would you like to proceed?" : "How would you like to proceed?",
-        choices: [Action.CANCEL, Action.MERGE, Action.RESET],
-      });
+    localChanges = withoutUnnecessaryChanges({ changes: localChanges, existing: gadgetHashes });
+    gadgetChanges = withoutUnnecessaryChanges({ changes: gadgetChanges, existing: localHashes });
```

Comment:
Don't we want to remove unnecessary changes before generating the conflicts between them?

Reply:
The end result shouldn't change, but yes we could do this before checking for conflicts.

In fact, we could do this inside of `getChanges` if we made it take `existing`:

```ts
let localChanges = getChanges({ from: filesVersionHashes, to: localHashes, existing: gadetHashes, ignore: [".gadget/"] });
let gadgetChanges = getChanges({ from: filesVersionHashes, to: gadgetHashes, existing: localHashes });
```

What do you think?

Reply:
Yeah I think moving it into `getChanges` is a good call.

Reply:
I was thinking that we wanted to drop unnecessary changes before asking the user about conflicts in case an unnecessary change was a conflict - but moving this into `getConflicts` achieves the same thing.

Reply:
Did this here: [Use withoutUnnecessaryChanges inside getChanges](https://github.com/gadget-inc/ggt/pull/1027/commits/81d0156b338291f9f1f02d6f2caec5a0de062133)